fast64_internal/f3d/f3d_node_gen.py

```diff
@@ -0,0 +1,972 @@
+    "outputs",
+    "dimensions",
+    "interface",
+    "internal_links",
+    "image_user",
+    "image",
+    "select",
+    "name",
+    "location_absolute",
+    "location",
+)
+EXCLUDE_FROM_GROUP_INPUT_OUTPUT = GENERAL_EXCLUDE + (
+    "bl_subtype_label",
+    "bl_socket_idname",
+    "display_shape",
+    "label",
+    "identifier",
+    "is_output",
+    "is_linked",
+    "is_multi_input",
+    "node",
+    "is_unavailable",
+    "show_expanded",
+    "link_limit",
+    "default_attribute_name",
+    "name",
+    "index",
+    "position",
+    "parent",
+    "socket_type",
+    "in_out",
+    "item_type",
+    "default_input",  # poorly documented, what does it do?
+)
+
+GLOBAL_DEFAULTS = {
+    "hide": False,
+    "mute": False,
+    "show_preview": False,
+    "label": "",
+    "description": "",
+    "parent": None,
+    "show_texture": False,
+    "use_custom_color": False,
+    "show_options": True,
+    "width": 100.0,
+    "width_hidden": 42.0,
+    "height": 100.0,
+    "text": None,
+    "hide_value": False,
+    "subtype": "NONE",
+    "attribute_domain": "POINT",
+    "clamp": False,
+    "max_value": 3.4028234663852886e38,
+    "min_value": -3.4028234663852886e38,
+    "pin_gizmo": False,
+    "warning_propagation": "ALL",
+    "is_inspect_output": False,
+    "factor_mode": "UNIFORM",
+    "clamp_factor": True,
+    "clamp_result": False,
+    "use_alpha": False,
+    "use_clamp": False,
+    "normalize": True,
+    "blend_type": "MIX",
+    "noise_type": "FBM",
+    "mode": "RGB",
+    "enabled": False,  # this may seem like an odd default, but in versions with added inputs those would be set to False
+    # unused in shader nodes
+    "hide_in_modifier": False,
+    "force_non_field": False,
+    "layer_selection_field": False,
+}
+
+
+class DefaultDefinition:
+    def __init__(self, names: list[str], defaults: dict[str, Any]):
+        self.names = names
+        self.defaults = GLOBAL_DEFAULTS | defaults
+
+
+DEFAULTS = [
+    DefaultDefinition(["NodeSocketFloat", "NodeSocketFloatFactor"], {"default_value": 0.0}),
+    DefaultDefinition(["NodeSocketInt"], {"default_value": 0}),
+    DefaultDefinition(["NodeSocketVector", "NodeSocketRotation"], {"default_value": (0.0, 0.0, 0.0)}),
+    DefaultDefinition(["NodeSocketColor"], {"default_value": (0.0, 0.0, 0.0, 0.0)}),
+    DefaultDefinition(["ShaderNodeMixRGB"], {"data_type": "RGBA"}),
+    DefaultDefinition(["NodeSocketVectorDirection"], {"subtype": "DIRECTION", "default_value": (0.0, 0.0, 0.0)}),
+]
+DEFAULTS = {name: definition.defaults for definition in DEFAULTS for name in definition.names}
+
+SCENE_PROPERTIES_VERSION = 2
+
+
+class ErrorState:
+    def __init__(self, error_message_queue: list[str] = None):
+        self.error_message_queue = error_message_queue or []
+        self.errors = []
+
+    def copy(self, message: str = None):
+        errors = ErrorState(self.error_message_queue.copy() + ([message] if message else []))
+        errors.errors = self.errors
```

Comment:
did you intend for error_message_queue to be a deepcopy but errors to only be shallow copied?

Reply:
yep